modules/subschema/src/main/scala/com.snowplowanalytics/iglu.schemaddl/jsonschema/subschema/package.scala

```diff
@@ -3,26 +3,56 @@ package com.snowplowanalytics.iglu.schemaddl.jsonschema.subschema
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.Schema
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.properties.CommonProperties._
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.properties.CommonProperties.Type._
-
 import dregex.Regex
 import io.circe.Json
 
+import scala.annotation.tailrec
+
 package object subschema {
 
   def isSubSchema(s1: Schema, s2: Schema): Compatibility =
     isSubType(simplify(canonicalize(s1)), simplify(canonicalize(s2)))
 
+  @tailrec
   def canonicalize(s: Schema): Schema =
     s match {
-      case s if s.`type`.contains(Null)    => s
-      case s if s.`type`.contains(Boolean) =>
-        if (s.`enum`.isDefined) s else s.copy(`enum` = Some(Enum(List(Json.True, Json.False))))
-      case s if s.`type`.contains(Integer) => s.copy(`type` = Some(Number))
-      case s if s.`type`.contains(Number)  => s
-      case s if s.`type`.contains(String)  => s
+      case s if s.`type`.contains(Null)                => s.copy(`enum` = None)
+      case s if s.`type`.contains(Boolean)             => canonicalizeBoolean(s)
+      case s if s.`type`.contains(Integer)             => s.copy(`type` = Some(Number))
+      case s if s.`type`.contains(Number)              => s
+      case s if s.`type`.contains(String)              => s
+      case s if s.`type`.isEmpty && s.`enum`.isDefined => canonicalize(canonicalizeEnum(s))
       case _ => s
     }
 
+  def canonicalizeBoolean(s: Schema): Schema =
+    if (s.`enum`.isDefined) s else s.copy(`enum` = Some(Enum(List(Json.True, Json.False))))
+
+  def canonicalizeEnum(s: Schema): Schema = {
+    val typeValue: Json => (Type, Option[Json]) =
+      _.fold(
+        jsonNull    = Null -> None,
+        jsonBoolean = v => Boolean -> Some(Json.fromBoolean(v)),
+        jsonNumber  = v => Number -> Some(Json.fromJsonNumber(v)),
+        jsonString  = v => String -> Some(Json.fromString(v)),
+        jsonArray   = v => Array -> Some(Json.fromValues(v)),
+        jsonObject  = v => Object -> Some(Json.fromJsonObject(v))
+      )
+
+    val splitByType: List[Json] => List[Schema] =
+      _.map(typeValue(_))
```

Comment:
analyze values of the enum, classifying them by type. Then collect them into separate schemas, one schema per type. 
The only special case is the Schema for the null value, which is returned - already canonicalised - with just type and no enum values